### ISSUE TITLE
hello,I am troubled with panorama ranging.Is there some way to get the relationship between panorama image and model?thank you

### ISSUE THREAD

**Issue by T-yanghui** [crs4/Deep3DLayout#2]

**Comment by giannipint** [comment]
Hello T-yanghui, yes, of course is possible. It is about the same process of features pooling from 3D coords to uv coords in image space. i share this quick and dirty sniplet of code:
def draw_projected_points(xyz, x_img, paint_img = True):
    
    Vt = torch.FloatTensor(xyz)

    uvd_verts = xyz2uvd(Vt)
       
    u, v, d = torch.unbind(uvd_verts.squeeze(0), dim=1) ###u,v are in norm image coords (-1,-1, top left)
          
    ###to pixels
    u = (u+1.0) * (x_img.shape[2]/2)
    v = (v+1.0) * (x_img.shape[1]/2)

    u = np.array(u, dtype=np.uint16)
    v = np.array(v, dtype=np.uint16)

    uv = np.stack((u, v), axis=-1)
       
    if(paint_img):
        dump = tools.x2image(x_img)##
        
    else:
        dump=np.zeros((x_img.shape[1],x_img.shape[2],3),np.uint8)

       
    for p in uv:
        dump = cv2.circle(dump, (p[0],p[1]), radius=3, color=(255, 0, 255), thickness=-1)

    ##print(dump.shape, dump.dtype)

    plt.figure(1111)
    plt.title('mesh to xyz2uvd')
    plt.imshow(cv2.UMat.get(dump))

To draw the 3D layout on the panorama image. The mapping between xyz and uv is given by: xyz2uvd

def xyz2uvd(xyz, eps = 0.001):
    x, y, z = torch.unbind(xyz, dim=2)
        
    x = x+eps
    y = y+eps
    z = z+eps
    
    u = torch.atan2(x, -y)
    v = - torch.atan(z / torch.sqrt(x**2 + y**2))
        
    pi = float(np.pi)##torch.acos(torch.zeros(1)).item() * 2

    u = u / pi
    v = (2.0 * v) / pi

    xx = x * x
    yy = y * y
    zz = z * z

    d = torch.sqrt(xx+yy+zz)

    u = torch.clamp(u, min=-1, max=1)
    v = torch.clamp(v, min=-1, max=1)
    d = torch.clamp(d, min=0, max=16)
        
    ###output: [batch_size x num_points x 3]

    output = torch.stack([u, v, d], dim=-1)
                
    return output

that is the same function included in gaf.py

Hoping it should help

**Comment by T-yanghui** [comment]
Hi giannipint,i am glad to hear your back and i have read it carefully. I just started leaning in this field,and taked two days to figure out how to run function draw_projected_points,but i am confused by the param xyz.And I also read GAF.py,but still dont understand the param.Can you tell me the exact meaning about the params or how to apply this function in the pipeline?
I want to find the relationship between panorama image pixel and output model,so that i can get  the distance between two points which on the panorama image.This is my practice direction during my student days.
Looking forward to your reply, and thanks again for your help.

**Comment by T-yanghui** [comment]
> Hello T-yanghui, yes, of course is possible. It is about the same process of features pooling from 3D coords to uv coords in image space. i share this quick and dirty sniplet of code: def draw_projected_points(xyz, x_img, paint_img = True):
> 
> ```
> Vt = torch.FloatTensor(xyz)
> 
> uvd_verts = xyz2uvd(Vt)
>    
> u, v, d = torch.unbind(uvd_verts.squeeze(0), dim=1) ###u,v are in norm image coords (-1,-1, top left)
>       
> ###to pixels
> u = (u+1.0) * (x_img.shape[2]/2)
> v = (v+1.0) * (x_img.shape[1]/2)
> 
> u = np.array(u, dtype=np.uint16)
> v = np.array(v, dtype=np.uint16)
> 
> uv = np.stack((u, v), axis=-1)
>    
> if(paint_img):
>     dump = tools.x2image(x_img)##
>     
> else:
>     dump=np.zeros((x_img.shape[1],x_img.shape[2],3),np.uint8)
> 
>    
> for p in uv:
>     dump = cv2.circle(dump, (p[0],p[1]), radius=3, color=(255, 0, 255), thickness=-1)
> 
> ##print(dump.shape, dump.dtype)
> 
> plt.figure(1111)
> plt.title('mesh to xyz2uvd')
> plt.imshow(cv2.UMat.get(dump))
> ```
> 
> To draw the 3D layout on the panorama image. The mapping between xyz and uv is given by: xyz2uvd
> 
> def xyz2uvd(xyz, eps = 0.001): x, y, z = torch.unbind(xyz, dim=2)
> 
> ```
> x = x+eps
> y = y+eps
> z = z+eps
> 
> u = torch.atan2(x, -y)
> v = - torch.atan(z / torch.sqrt(x**2 + y**2))
>     
> pi = float(np.pi)##torch.acos(torch.zeros(1)).item() * 2
> 
> u = u / pi
> v = (2.0 * v) / pi
> 
> xx = x * x
> yy = y * y
> zz = z * z
> 
> d = torch.sqrt(xx+yy+zz)
> 
> u = torch.clamp(u, min=-1, max=1)
> v = torch.clamp(v, min=-1, max=1)
> d = torch.clamp(d, min=0, max=16)
>     
> ###output: [batch_size x num_points x 3]
> 
> output = torch.stack([u, v, d], dim=-1)
>             
> return output
> ```
> 
> that is the same function included in gaf.py
> 
> Hoping it should help



> Hello T-yanghui, yes, of course is possible. It is about the same process of features pooling from 3D coords to uv coords in image space. i share this quick and dirty sniplet of code: def draw_projected_points(xyz, x_img, paint_img = True):
> 
> ```
> Vt = torch.FloatTensor(xyz)
> 
> uvd_verts = xyz2uvd(Vt)
>    
> u, v, d = torch.unbind(uvd_verts.squeeze(0), dim=1) ###u,v are in norm image coords (-1,-1, top left)
>       
> ###to pixels
> u = (u+1.0) * (x_img.shape[2]/2)
> v = (v+1.0) * (x_img.shape[1]/2)
> 
> u = np.array(u, dtype=np.uint16)
> v = np.array(v, dtype=np.uint16)
> 
> uv = np.stack((u, v), axis=-1)
>    
> if(paint_img):
>     dump = tools.x2image(x_img)##
>     
> else:
>     dump=np.zeros((x_img.shape[1],x_img.shape[2],3),np.uint8)
> 
>    
> for p in uv:
>     dump = cv2.circle(dump, (p[0],p[1]), radius=3, color=(255, 0, 255), thickness=-1)
> 
> ##print(dump.shape, dump.dtype)
> 
> plt.figure(1111)
> plt.title('mesh to xyz2uvd')
> plt.imshow(cv2.UMat.get(dump))
> ```
> 
> To draw the 3D layout on the panorama image. The mapping between xyz and uv is given by: xyz2uvd
> 
> def xyz2uvd(xyz, eps = 0.001): x, y, z = torch.unbind(xyz, dim=2)
> 
> ```
> x = x+eps
> y = y+eps
> z = z+eps
> 
> u = torch.atan2(x, -y)
> v = - torch.atan(z / torch.sqrt(x**2 + y**2))
>     
> pi = float(np.pi)##torch.acos(torch.zeros(1)).item() * 2
> 
> u = u / pi
> v = (2.0 * v) / pi
> 
> xx = x * x
> yy = y * y
> zz = z * z
> 
> d = torch.sqrt(xx+yy+zz)
> 
> u = torch.clamp(u, min=-1, max=1)
> v = torch.clamp(v, min=-1, max=1)
> d = torch.clamp(d, min=0, max=16)
>     
> ###output: [batch_size x num_points x 3]
> 
> output = torch.stack([u, v, d], dim=-1)
>             
> return output
> ```
> 
> that is the same function included in gaf.py
> 
> Hoping it should help
Hi giannipint,i am glad to hear your back and i have read it carefully. I just started leaning in this field,and taked two days to figure out how to run function draw_projected_points,but i am confused by the param xyz.And I also read GAF.py,but still dont understand the param.Can you tell me the exact meaning about the params or how to apply this function in the pipeline?
I want to find the relationship between panorama image pixel and output model,so that i can get the distance between two points which on the panorama image.This is my practice direction during my student days.
Looking forward to your reply, and thanks again for your help.
